### PR TITLE
Implement LDAP support

### PR DESCRIPTION
**Pull request by lindskogen**
Disable change password when ldap is enabled

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -189,28 +227,31 @@ function auth(data) {
...
-				if (data.token === client.token) {
-					success = true;
-				}
-			} else if (client.config.user === data.user) {
-				if (bcrypt.compareSync(data.password || "", client.config.password)) {
-					success = true;
-				}
-			}
+		client = manager.findClient(data.user, data.token);
+		var signedIn = data.token && client.token === data.token;
```
Comment (Dr-Horv) [inline src/server.js]: Should probably check that `client` exist before trying to access token. I think  `findClient` can return false/undefined.

**Comment by xPaw** [comment]
What should happen if you have public mode and ldap enabled at the same time?

**Comment by lindskogen** [comment]
If public mode is enabled the server won't even try to authenticate the user, so it won't be a problem. Maybe you're thinking more along the lines showing an error message and exiting?

**Comment by xPaw** [comment]
Yeah, showing a message would probably be helpful to avoid issues.

**Comment by lindskogen** [comment]
@xPaw ok, rebased again :smile:

**Comment by maxpoulin64** [comment]
From a code standpoint, this looks fine to me. I would test that, but I have a bit of trouble getting around to understand that mess. It' a massive pain in the ass to set up and I can't find any tutorial that doesn't assume one is already an LDAP expert or any tutorial that explains how to set up a minimal OpenLDAP server that doesn't involve a bazillion config files, files and databases everywhere. So I'll pass on the testing of this one.

**Comment by YaManicKill** [comment]
So, this would probably be a good thing to get in, but I guess none of @thelounge/maintainers are happy enough tesitng ldap stuff.

I propose that we test the PR (once @lindskogen rebases again, please :-P) and ensure there are no regressions in the other user stuff, and if not...then we merge in. Assuming people are happy with the code.

This is something we have been asked for quite a bit, so we can't just ignore it.

So either we take this option, or someone learns ldap and delves into the horribleness. I used to have an ldap server I could use, but I never set it up, and I don't have access to it anymore. I wouldn't know where to start.

**Comment by YaManicKill** [comment]
Or, and here's a better solution, https://jumpcloud.com/daas-product/index

Free for 10 user hosted ldap solution.

**Comment by jedahan** [comment]
jumpcloud looks awesome for testing. If someone does want to dig into self-hosted ldap testing, the easiest looking stuff I've found has been https://github.com/osixia/docker-openldap for the server and https://github.com/osixia/docker-phpLDAPadmin for admin gui

**Comment by jedahan** [comment]
And the build-from-scratch possibly-addable-to-npm-test-suite option: http://ldapjs.org/examples.html

**Comment by jedahan** [comment]
I ran the patches today, had to fix up some of the merge conflicts, and it works against our production ldap pretty flawlessly. Would love to see this merged in soon.

**Comment by astorije** [comment]
Thanks for the update, @jedahan, that is great news indeed!

@lindskogen, could you rebase your PR on the `master` branch and fix the conflict, please? @jedahan, if you had hard time resolving the conflicts, maybe it's worth sharing what you did :-)

File: package.json [hunk 2]
```diff
@@ -32,6 +32,7 @@
     "commander": "2.9.0",
     "event-stream": "3.3.2",
     "express": "4.13.4",
+    "ldapjs": "~1.0.0",
```
Comment (astorije) [inline package.json]: Could you use `"1.0.0"` instead, please?

**Comment by jedahan** [comment]
The merge conflicts were minimal, though I think its a tiny bit buggy, I can share my changes tomorrow at work. It was blindly copy-pasting until it worked.

**Comment by lindskogen** [comment]
I actually tried resolving the conflicts last week but got stuck on webIRC stuff which I know nothing about 😩

**Comment by astorije** [comment]
@maxpoulin64, do you think you could help @lindskogen resolving conflicts around WEBIRC? Maybe you guys could sync on the IRC channel (`#thelounge`)?

**Comment by maxpoulin64** [comment]
@astorije, @lindskogen: They don't actually conflict. It's just that things have moved a lot: I just tried fixing it and the solution is basically to just manually apply the patch instead of trying to resolve merge conflicts. You don't need to touch the WebIRC stuff, just leave it there and make sure it still runs (it applies to both private and public modes).

I can do it if you still have trouble with it.

**Comment by thisisdarshan** [comment]
Created PR#477 that resolves all the conflict.

**Comment by lindskogen** [comment]
Closing this in favor of #477

**Pull request closed without merging**